File: backend/app/schemas/dispatch.py
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel
from typing import Literal

from app.models.dispatch import DispatchStatus
from app.models.security_officer import AvailabilityStatus

class DispatchCandidateRes(BaseModel):
    """One ranked candidate row for a given alert"""
    id: UUID
    alert_id: UUID
    officer_id: UUID | None
    rank: int | None
    score: float | None
    distance: float | None
    eta: float | None
    workload: int | None
    status: DispatchStatus
    officer_availability: AvailabilityStatus | None
    is_location_stale: bool
    created_at: datetime
    notified_at: datetime | None
    responded_at: datetime | None

File: frontend/src/lib/api/dispatch.ts
import { apiFetch } from "./alert";

export type DispatchStatus =
  | "SELECTED"
  | "PENDING"
  | "QUEUED"
  | "NOTIFIED"
  | "ACCEPTED"
  | "DECLINED"
  | "TIMED_OUT"
  | "NO_CANDIDATE";

export type DispatchAction = "ACCEPT" | "DECLINE";

export interface DispatchCandidate {
  id: string;
  alert_id: string;
  officer_id: string | null;
  rank: number | null;
  score: number | null;
  distance: number | null;
  eta: number | null;
  workload: number | null;
  status: DispatchStatus;
  officer_availability: string | null;
  is_location_stale: boolean;
  created_at: string;
  notified_at: string | null;
  responded_at: string | null;
}

interface RespondDispatchResponse {
  status: number;
  message: string | null;
  data: DispatchCandidate | null;
}

export async function respondToDispatch(
  dispatchId: string,
  action: DispatchAction,
): Promise<DispatchCandidate | null> {
  const response = await apiFetch<RespondDispatchResponse>(
    `/dispatch/${dispatchId}/respond`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ action }),
    },
  );

  return response.data;
}
